src/c++/perf_analyzer/test_command_line_parser.cc

```diff
@@ -0,0 +1,1040 @@
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(!parser.UsageCalled());
+
+      exp->max_threads = 1;
+      exp->max_threads_specified = true;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("set to max")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--max-threads", "65535"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(!parser.UsageCalled());
+
+      exp->max_threads = 65535;
+      exp->max_threads_specified = true;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("missing value")
+    {
+      int argc = 4;
+      char* argv[argc] = {app_name, "-m", model_name, "--max-threads"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(parser.UsageCalled());
+
+      // NOTE: Empty message is not helpful
+      //
+      CHECK_STRING("Usage Message", parser.GetUsageMessage(), "");
+      // BUG: Dumping string "option '--max-threads' requires an argument"
+      // directly to std::out, instead of through usage()
+      //
+
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("bad value")
+    {
+      int argc = 4;
+      char* argv[argc] = {app_name, "-m", model_name, "--max-threads", "bad"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      REQUIRE(parser.UsageCalled());
+
+      // NOTE: Empty message is not helpful
+      //
+      CHECK_STRING("Usage Message", parser.GetUsageMessage(), "");
+      // BUG: Dumping string "option '--max-threads' requires an argument"
+      // directly to std::out, instead of through usage()
+      //
+
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+  }
+
+  SUBCASE("Option : --sequence-length")
+  {
+    SUBCASE("set to 2000")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--sequence-length",
+                          "2000"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(!parser.UsageCalled());
+
+      exp->sequence_length = 2000;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+  }
+
+  SUBCASE("Option : --percentile")
+  {
+    SUBCASE("set to 25")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--percentile", "25"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(!parser.UsageCalled());
+
+      exp->percentile = 25;
+      CHECK_PARAMS(act, exp);
+      optind = 1;
+    }
+
+    SUBCASE("set to 225 - overflow check")
+    {
+      int argc = 5;
+      char* argv[argc] = {app_name, "-m", model_name, "--percentile", "225"};
+
+      REQUIRE_NOTHROW(act = parser.Parse(argc, argv));
+      CHECK(parser.UsageCalled());
+      CHECK_STRING(
+          "Usage Message", parser.GetUsageMessage(),
+          "percentile must be -1 for not reporting or in range (0, 100)");
+
+      exp->percentile = 225;
+      CHECK_PARAMS(act, exp);
```

Comment:
A call to `Usage()` always results in exiting. So what's the value-add of testing the params struct after `Usage()` gets called (here and elsewhere)?

Reply:
Because that is the current behavior of the class. From the classes perspective, it does not know the application is being exited. That is a business decision higher in the code. In the future, perhaps that is not the case. However, the current application IS setting those values. That is the behavior and the test captures it. If, in the future we decide that the behavior is to NOT modify those values, then we can clear this field and run to make sure nothing is changed from default.

Note this type of unit testing, those that capture existing behavior, must try to catch any and all changes.

Reply:
Makes sense. Thanks for explaining.